tss-esapi/examples/hmac.rs

```diff
@@ -0,0 +1,151 @@
+        // knows it's content.
+        .with_sensitive_data_origin(true)
+        // This key requires "authentication" to the TPM to access - this can be
+        // an HMAC or password session. HMAC sessions are used by default with
+        // the "execute_with_nullauth_session" function.
+        .with_user_with_auth(true)
+        // This key has the ability to decrypt
+        .with_decrypt(true)
+        // This key may only be used to encrypt or sign objects that are within
+        // the TPM - it can not encrypt or sign external data.
+        .with_restricted(true)
+        .build()
+        .expect("Failed to build object attributes");
+
+    let primary_pub = PublicBuilder::new()
+        // This key is a symmetric key.
+        .with_public_algorithm(PublicAlgorithm::SymCipher)
+        .with_name_hashing_algorithm(HashingAlgorithm::Sha256)
+        .with_object_attributes(object_attributes)
+        .with_symmetric_cipher_parameters(
+            SymmetricCipherParameters::new(SymmetricDefinitionObject::AES_128_CFB)
+        )
+        .with_symmetric_cipher_unique_identifier(Digest::default())
+        .build()
+        .unwrap();
+
+    let primary = context.execute_with_nullauth_session(|ctx| {
+        // Create the key under the "owner" hierachy. Other hierachies are platform
+        // which is for boot services, null which is ephemeral and resets after a reboot,
+        // and endorsement which allows key certification.
+        ctx.create_primary(Hierarchy::Owner, primary_pub, None, None, None, None)
+    }).unwrap();
+
+    // Create the HMAC key. This key exists under the primary key in it's hierarchy
+    // and can only be used if the same primary key is recreated from the parameters
+    // defined above.
+
+    let object_attributes = ObjectAttributesBuilder::new()
+        .with_fixed_tpm(true)
+        .with_fixed_parent(true)
+        .with_st_clear(false)
+        .with_sensitive_data_origin(true)
+        .with_user_with_auth(true)
+        // The key is used only for signing.
+        .with_sign_encrypt(true)
+        .build()
+        .expect("Failed to build object attributes");
+
+    let key_pub = PublicBuilder::new()
+        // This key is a HMAC key
+        .with_public_algorithm(PublicAlgorithm::KeyedHash)
+        .with_name_hashing_algorithm(HashingAlgorithm::Sha256)
+        .with_object_attributes(object_attributes)
+        .with_keyed_hash_parameters(PublicKeyedHashParameters::new(KeyedHashScheme::HMAC_SHA_256))
+        .with_keyed_hash_unique_identifier(Digest::default())
+        .build()
+        .unwrap();
+
+    let (enc_private, public) = context.execute_with_nullauth_session(|ctx| {
+        // Create the HMAC key given our primary key as it's parent. This returns the private
+        // and public portions of the key. It's *important* to note that the private component
+        // is *encrypted* by the primary key. It is not plaintext or leaked in this step.
```

Comment:
The `by the primary key` bit on this line might be a bit confusing, because the encryption is done with a key _associated with_ the primary/parent key.

Reply:
I've read books on TPMs, read parts of the spec, and still managed to not know how this works. I think TPM's really do have an accessibility/communication issue 😅
Anyway thanks for the correction, I've fixed the comment.

Reply:
Agree, going through TPM docs is a Kafkaesque endeavour...

Reply:
Yep. ... It's why I'm hoping some extra examples here of common and practical use cases will help people :)